### PR TITLE
test!: add unit test

### PR DESCRIPTION
**Pull request by mandrake-sw**
- Added unit tests
- Fixed port reading [issue](https://github.com/Supergiovane/KNXUltimate/pull/36/commits/f3498b8dc22551410428af671e7e29d91170cb68#diff-6ff63d6c78aa4d6f849d599f7fa9da59e7f6eb1f3959a574788013a20381ba08R105)
- Bump minimum Node.js required version to 16

BREAKING CHANGE: Bump minimum Node.js required version to 16

### CODE REVIEW

**Comment by robertsLando** [comment]
@mandrake-sw Add node 22 to CI

File: test/unit/utils.test.ts [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+		})
+
+		it('should handle negative numbers', () => {
+			assert.strictEqual(round(-1.2345, 2), -1.23)
+			assert.strictEqual(round(-1.2365, 2), -1.24)
+		})
+
+		it('should handle zero', () => {
+			assert.strictEqual(round(0, 2), 0)
+		})
```
Comment (robertsLando) [inline test/unit/utils.test.ts:197]: add a case here that covers possible js round errors. See https://www.robinwieruch.de/javascript-rounding-errors/

Comment (mandrake-sw) [reply test/unit/utils.test.ts:197]: Done, used Number.EPSILON

File: test/utils/MockKNXServer.ts [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+		this.expectedTelegrams = capturedTelegrams
+		this.socket = socket
+
+		// For UDP sockets
+		if (socket instanceof UDPSocket) {
+			console.log('[MOCK] Setting up UDP mock server')
+			socket.on('message', (msg: Buffer, rinfo: any) => {
+				console.log('[MOCK] UDP message received:', msg.toString('hex'))
+				this.handleRequest(msg, rinfo)
+			})
```
Comment (robertsLando) [inline test/utils/MockKNXServer.ts:26]: this creates an infinite loop, use previous way

File: test/utils/MockKNXServer.ts [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+		this.setupRequestHandlers()
+
+		console.log(
+			'[MOCK] MockKNXServer initialized with handlers:',
+			Array.from(this.requestHandlers.keys()),
+		)
+	}
+
+	private setupRequestHandlers() {
+		this.expectedTelegrams.forEach((telegram) => {
```
Comment (robertsLando) [inline test/utils/MockKNXServer.ts:46]: You should remove the request from requestHandlers once you handled it

File: test/utils/MockKNXServer.ts [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+		})
+	}
+
+	private handleRequest(data: Buffer, rinfo?: any) {
+		const requestHex = data.toString('hex')
+		console.log('[MOCK] Processing request:', requestHex)
+
+		const handler = this.requestHandlers.get(requestHex)
+		if (handler) {
+			console.log('[MOCK] Handler found, executing')
```
Comment (robertsLando) [inline test/utils/MockKNXServer.ts:91]: Use KnxLoG instead of console.log (everywhere)

File: test/integration/KNXClient.test.ts [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+	},
+]
+
+describe('KNXClient Tests', () => {
+	test(
+		'should discover KNX interfaces',
+		{ timeout: TEST_TIMEOUT },
+		async () => {
+			console.log('[TEST] Starting discovery test')
+			console.log('[TEST] Running in CI:', process.env.CI === 'true')
```
Comment (robertsLando) [inline test/integration/KNXClient.test.ts:27]: remove all console.log

**Review by robertsLando — CHANGES_REQUESTED** [review]

**Comment by Supergiovane** [comment]
As a side note about logging, we'll use [Winston](https://github.com/winstonjs/winston/tree/master) as a definitive logger.
I've tried some other logger, but Winston is more effective and widely supported.
Nothing changes yet, it all remains the same, so no worry. The switch between one another will be seamlessly transparent.

File: test/utils/MockKNXServer.ts [hunk 6]
```diff
@@ -28,27 +22,25 @@ export default class MockKNXServer {
...
+			}.bind(this)
 		} else {
 			console.log('[MOCK] UDP socket detected')
 			const originalSend = socket.send
-			socket.send = (data: Buffer, ...args: any[]) => {
+			socket.send = function (data: Buffer, ...args: any[]) {
 				this.onRequest(data)
 				return originalSend.call(socket, data, ...args)
-			}
+			}.bind(this)
```
Comment (robertsLando) [inline test/utils/MockKNXServer.ts:35]: useless bind

**Review by robertsLando — APPROVED** [review]

File: test/utils/MockKNXServer.ts [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+				this.onRequest(data)
+				return originalWrite.call(socket, data)
+			}.bind(this)
+		} else {
+			console.log('[MOCK] UDP socket detected')
+			const originalSend = socket.send
+			socket.send = function (data: Buffer, ...args: any[]) {
+				this.onRequest(data)
+				return originalSend.call(socket, data, ...args)
+			}.bind(this)
```
Comment (robertsLando) [inline test/utils/MockKNXServer.ts:35]: keep arrow function here and remove the bind

File: test/dtp/dpt1.test.ts [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+			assert.strictEqual(alarmType.use, 'FB')
+		})
+
+		test('should have correct window/door (019) subtype', () => {
+			const windowDoorType = DPT1.subtypes['019']
+			assert.strictEqual(windowDoorType.name, 'Window/Door')
+			assert.deepEqual(windowDoorType.enc, { 0: 'closed', 1: 'open' })
+			assert.strictEqual(windowDoorType.use, 'G')
+		})
+	})
```
Comment (robertsLando) [inline test/dtp/dpt1.test.ts:76]: I think this is not necessary

File: test/dtp/dpt2.test.ts [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+					`Subtype ${id} should have a use property`,
+				)
+				assert.ok(
+					subtype.desc,
+					`Subtype ${id} should have a description`,
+				)
+				assert.ok(subtype.enc, `Subtype ${id} should have encodings`)
+			})
+		})
+	})
```
Comment (robertsLando) [inline test/dtp/dpt2.test.ts:122]: Not necessary

File: test/dtp/dpt3.test.ts [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+		test('should verify all subtypes have required properties', () => {
+			Object.entries(DPT3.subtypes).forEach(([id, subtype]) => {
+				assert.ok(subtype.name, `Subtype ${id} should have a name`)
+				assert.ok(
+					subtype.desc,
+					`Subtype ${id} should have a description`,
+				)
+			})
+		})
+	})
```
Comment (robertsLando) [inline test/dtp/dpt3.test.ts:123]: not necessary

**Review by robertsLando — CHANGES_REQUESTED** [review]
I think that once done with all dpts we can merge this

File: src/KNXClient.ts [hunk 11]
```diff
@@ -611,6 +651,14 @@ export default class KNXClient extends TypedEventEmitter<KNXClientEventCallbacks
 		)
 	}
 
+	public getSniffingBuffers() {
+		return this.sniffingPackets
+	}
+
+	public clearSniffingBuffers(): void {
+		this.sniffingPackets = []
+	}
```
Comment (robertsLando) [inline src/KNXClient.ts:660]: not needed anymore

**Review by robertsLando — CHANGES_REQUESTED** [review]

**Comment by Supergiovane** [comment]
> I think that once done with all dpts we can merge this

Lemme know when ready and i'll squash/merge

**Comment by robertsLando** [comment]
@mandrake-sw Remember also to add coverage report to tests

**Comment by robertsLando** [comment]
Test folder should not be included in the test report.

Errors unit tests are a bit overkill, it would be better to test situations where those errors are emitted and reproduce that.

DPT library tests have a test with a comment `continue for all dpts`, I would drop that test as it's a bit useless as you  should list a lot of dpts there

File: test/integration/KNXClient.test.ts [hunk 12]
```diff
@@ -0,0 +1,473 @@
...
+	]
+}
+
+const getMockToggleResponses = (): SnifferPacket[] => {
+	const localIp = getDefaultIpLocal()
+	const knxGwIp = ipToHex(MockKNXServer.host)
+
+	if (!localIp) {
+		throw new Error('No local IP found')
+	}
```
Comment (robertsLando) [inline test/integration/KNXClient.test.ts:64]: This doesn't need to be a function, just declare a var with the packets

File: test/integration/KNXClient.test.ts [hunk 13]
```diff
@@ -0,0 +1,473 @@
...
+			request: '06100209001051000801000000000000',
+			deltaReq: 1250,
+			response: '0610020a00085100',
+			resType: 'KNXDisconnectResponse',
+			deltaRes: 4,
+		},
+	]
+}
+
+const getMockPacketsForDisconnectTest = (): SnifferPacket[] => {
```
Comment (robertsLando) [inline test/integration/KNXClient.test.ts:149]: same

File: test/integration/KNXClient.test.ts [hunk 14]
```diff
@@ -0,0 +1,473 @@
...
+					)
+
+					// Override error log to fail test when no matching response found
+					const originalError = server['error'].bind(server)
+					server['error'] = (message: string) => {
+						if (message.includes('No matching response found')) {
+							throw new Error(`MockKNXServer error: ${message}`)
+						}
+						originalError(message)
+					}
```
Comment (robertsLando) [inline test/integration/KNXClient.test.ts:365]: change approach instead, emit an event when there is an error and intercept it here with server.on('error'). Remember that in order to do this mocksewrver should extend eventemitter and call super() on constructor. See how I did it on KnxClient

**Review by robertsLando — CHANGES_REQUESTED** [review]

**Pull request merged**